File: webapp/static/js/prompts.js
// ATLAS Playground — prompts, @seat, A/B, evals. Forms only, no prompt().
// Runs measure through the LINE's play package; versions fold whole;
// evals score exact-match; the model is always a measurement.
const Play = {
  current: null,

  async render(el) {
    el.innerHTML = `
      <div class="page-header"><div><div class="page-title">Playground</div>
      <div class="page-subtitle">Versioned prompts, single-seat asks, honest compares</div></div></div>
      <div class="grid-2">
        <div class="card"><div class="card-title">Registry</div>
          <div id="play-list"><div class="loading">Loading...</div></div>
          <div class="card-title mt-16">Editor</div>
          <input id="play-name" type="text" placeholder="prompt name (lowercase, - _)" />
          <input id="play-desc" type="text" placeholder="description" />
          <textarea id="play-body" rows="8" placeholder="Body with {{vars}}..."></textarea>
          <button class="btn btn-sm" id="play-save">Save new version</button>
          <div id="play-save-status" class="muted"></div>
        </div>
        <div class="card"><div class="card-title">Run</div>
          <input id="play-vars" type="text" placeholder='vars JSON, e.g. {"word":"hello"}' value="{}" />
          <div class="flex"><input id="play-voice" type="text" placeholder="voice (empty = default route)" />
          <input id="play-version" type="number" min="0" placeholder="version (0 = latest)" value="0" /></div>
          <button class="btn btn-sm" id="play-run">Run</button>
          <div id="play-out"></div>
          <div class="card-title mt-16">A/B compare</div>
          <div class="flex"><input id="play-vera" type="number" min="1" value="1" />
          <input id="play-verb" type="number" min="1" value="2" />
          <button class="btn btn-sm" id="play-compare">Compare</button></div>
          <div id="play-cmp"></div>
        </div>
      </div>
      <div class="grid-2 mt-16">
        <div class="card"><div class="card-title">@seat — one seat, once</div>
          <input id="seat-q" type="text" placeholder="@manjuel what holds" />
          <div class="flex"><input id="seat-voice" type="text" placeholder="model override (measurement)" />
          <input id="seat-method" type="text" placeholder="method (one run, then spent)" /></div>
          <button class="btn btn-sm" id="seat-ask">Ask the seat</button>
          <div id="seat-out"></div>
        </div>
        <div class="card"><div class="card-title">Eval — score over a dataset</div>
          <div class="flex"><input id="eval-dataset" type="text" placeholder="dataset (evals/name.json)" />
          <button class="btn btn-sm" id="eval-run">Run eval</button></div>
          <div id="eval-out"></div>
        </div>
      </div>`;
    document.getElementById('play-save').onclick = () => this.save();
    document.getElementById('play-run').onclick = () => this.run();
    document.getElementById('play-compare').onclick = () => this.compare();
    document.getElementById('seat-ask').onclick = () => this.seat();
    document.getElementById('eval-run').onclick = () => this.eval();
    await this.refresh();
  },

  async refresh() {
    const box = document.getElementById('play-list');
    try {
      const r = await API.listPrompts();
      const names = [];
      for (const line of (r.prompts || '').split('\n')) {
        const m = /^\s*-\s(\S+)\s+v(\d+)\s+\(kept:\s*(\[[^\]]*\])\)\s*(?:—\s*(.*))?/.exec(line);
        if (m) names.push({ name: m[1], ver: +m[2], kept: m[3], desc: m[4] || '' });
      }
      box.innerHTML = names.length ? names.map(p =>
        `<div class="chat-session" data-n="${escHtml(p.name)}"><b>${escHtml(p.name)}</b> v${p.ver} <span class="muted">${escHtml(p.desc)}</span></div>`
      ).join('') : '<div class="empty-text">No prompts yet. Save the first below.</div>';
      box.querySelectorAll('.chat-session').forEach(d => {
        d.onclick = () => this.open(d.dataset.n);
      });
    } catch { box.innerHTML = '<div class="empty-text">MCP unreachable.</div>'; }
  },

  async open(name) {
    this.current = name;
    document.getElementById('play-name').value = name;
    try {
      const r = await API.getPrompt(name, 0);
      const body = (r.prompt || '').split('\n\n').slice(1).join('\n\n');
      document.getElementById('play-body').value = body;
    } catch { toast('Could not open prompt', 'error'); }
  },

  async save() {
    const name = document.getElementById('play-name').value.trim();
    const body = document.getElementById('play-body').value;
    const desc = document.getElementById('play-desc').value;
    try {
      const r = await API.savePrompt(name, body, desc);
      document.getElementById('play-save-status').textContent = r.text;
      this.current = name;
      await this.refresh();
    } catch (e) { document.getElementById('play-save-status').textContent = 'Refused: ' + e.message; }
  },

  name() { return this.current || document.getElementById('play-name').value.trim(); },

  async run() {
    const out = document.getElementById('play-out');
    out.innerHTML = '<div class="loading">Measuring...</div>';
    try {
      const r = await API.runPrompt(this.name(),
        document.getElementById('play-vars').value,
        +document.getElementById('play-version').value || 0,
        document.getElementById('play-voice').value);
      out.innerHTML = `<pre>${escHtml(r.text)}</pre>`;
    } catch (e) { out.innerHTML = `<div class="empty-text">Refused: ${escHtml(e.message)}</div>`; }
  },

  async compare() {
    const box = document.getElementById('play-cmp');
    box.innerHTML = '<div class="loading">Running both versions...</div>';
    try {
      const r = await API.comparePrompts(this.name(),
        +document.getElementById('play-vera').value || 1,
        +document.getElementById('play-verb').value || 2,
        document.getElementById('play-vars').value,
        document.getElementById('play-voice').value);
      box.innerHTML = `<pre>${escHtml(r.text)}</pre>`;
    } catch (e) { box.innerHTML = `<div class="empty-text">Refused: ${escHtml(e.message)}</div>`; }
  },

  async seat() {
    const out = document.getElementById('seat-out');
    out.innerHTML = '<div class="loading">Asking...</div>';
    try {
      const r = await API.seatAsk(
        document.getElementById('seat-q').value,
        '', document.getElementById('seat-voice').value,
        document.getElementById('seat-method').value);
      out.innerHTML = `<pre>${escHtml(r.text)}</pre>`;
    } catch (e) { out.innerHTML = `<div class="empty-text">Refused: ${escHtml(e.message)}</div>`; }
  },

  async eval() {
    const out = document.getElementById('eval-out');
    out.innerHTML = '<div class="loading">Scoring...</div>';
    try {
      const r = await API.evalPrompt(this.name(),
        document.getElementById('eval-dataset').value,
        0, document.getElementById('play-voice').value);
      const m = /(\d+)\/(\d+) pass/.exec(r.text || '');
      const rate = m ? (+m[1] / +m[2]) : 0;
      out.innerHTML = `<pre>${escHtml(r.text)}</pre>
        <button class="btn btn-sm" id="eval-record">Record score as eval</button>`;
      document.getElementById('eval-record').onclick = async () => {
        try {
          await API.addEval({ name: 'prompt-eval:' + Play.name(), score: rate, passed: rate === 1, detail: r.text });
          toast('Score recorded');
        } catch { toast('Could not record', 'error'); }
      };
    } catch (e) { out.innerHTML = `<div class="empty-text">Refused: ${escHtml(e.message)}</div>`; }
  }
};

File: webapp/handlers/prompts.go
// Prompt faces (N4): registry, measured runs, compares and evals — all
// through MCP /rpc. The webapp owns no playground logic; versions fold,
// runs measure and evals score in the LINE's play package. Whole answers
// only: compares and evals need complete outputs to judge honestly.
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"
	"strings"
)

func qVersion(r *http.Request, key string) float64 {
	n, _ := strconv.Atoi(r.URL.Query().Get(key))
	return float64(n)
}

// ListPrompts proxies prompt_list.
func (h *Handlers) ListPrompts(w http.ResponseWriter, r *http.Request) {
	text, err := h.rpcCallAs(r, "prompt_list", map[string]any{})
	if err != nil {
		jsonErr(w, 502, err.Error())
		return
	}
	jsonResp(w, map[string]string{"prompts": text})
}

// GetPrompt proxies prompt_get.
func (h *Handlers) GetPrompt(w http.ResponseWriter, r *http.Request) {
	text, err := h.rpcCallAs(r, "prompt_get", map[string]any{
		"name": r.URL.Query().Get("name"), "version": qVersion(r, "version")})
	if err != nil {
		jsonErr(w, 502, err.Error())
		return
	}
	jsonResp(w, map[string]string{"prompt": text})
}

// SavePrompt proxies prompt_save and announces the fold.
func (h *Handlers) SavePrompt(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name        string `json:"name"`
		Body        string `json:"body"`
		Description string `json:"description"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		jsonErr(w, 400, "invalid json")
		return
	}
	if strings.TrimSpace(body.Name) == "" || strings.TrimSpace(body.Body) == "" {
		jsonErr(w, 400, "name and body are required")
		return
	}
	text, err := h.rpcCallAs(r, "prompt_save", map[string]any{
		"name": body.Name, "body": body.Body, "description": body.Description})
	if err != nil {
		jsonErr(w, 502, err.Error())
		return
	}
	h.broadcast("prompt_saved", map[string]string{"text": text})
	jsonResp(w, map[string]string{"status": "ok", "text": text})
}

// RunPrompt proxies prompt_run and announces the measurement.
func (h *Handlers) RunPrompt(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name    string  `json:"name"`
		Vars    string  `json:"vars"`
		Version float64 `json:"version"`
		Voice   string  `json:"voice"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		jsonErr(w, 400, "invalid json")
		return
	}
	if strings.TrimSpace(body.Name) == "" {
		jsonErr(w, 400, "name is required")
		return
	}
	text, err := h.rpcCallAs(r, "prompt_run", map[string]any{
		"name": body.Name, "vars": body.Vars,
		"version": body.Version, "voice": body.Voice})
	if err != nil {
		jsonErr(w, 502, err.Error())
		return
	}
	h.broadcast("prompt_ran", map[string]string{"text": text})
	jsonResp(w, map[string]string{"status": "ok", "text": text})
}

// ComparePrompts proxies prompt_compare.
func (h *Handlers) ComparePrompts(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name  string  `json:"name"`
		VerA  float64 `json:"vera"`
		VerB  float64 `json:"verb"`
		Vars  string  `json:"vars"`
		Voice string  `json:"voice"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		jsonErr(w, 400, "invalid json")
		return
	}
	if strings.TrimSpace(body.Name) == "" {
		jsonErr(w, 400, "name is required")
		return
	}
	text, err := h.rpcCallAs(r, "prompt_compare", map[string]any{
		"name": body.Name, "vera": body.VerA, "verb": body.VerB,
		"vars": body.Vars, "voice": body.Voice})
	if err != nil {
		jsonErr(w, 502, err.Error())
		return
	}
	h.broadcast("prompt_compared", map[string]string{"text": text})
	jsonResp(w, map[string]string{"status": "ok", "text": text})
}

// EvalPrompt proxies prompt_eval and announces the score.
func (h *Handlers) EvalPrompt(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name    string  `json:"name"`
		Dataset string  `json:"dataset"`
		Version float64 `json:"version"`
		Voice   string  `json:"voice"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		jsonErr(w, 400, "invalid json")
		return
	}
	if strings.TrimSpace(body.Name) == "" || strings.TrimSpace(body.Dataset) == "" {
		jsonErr(w, 400, "name and dataset are required")
		return
	}
	text, err := h.rpcCallAs(r, "prompt_eval", map[string]any{
		"name": body.Name, "dataset": body.Dataset,
		"version": body.Version, "voice": body.Voice})
	if err != nil {
		jsonErr(w, 502, err.Error())
		return
	}
	h.broadcast("prompt_evaled", map[string]string{"text": text})
	jsonResp(w, map[string]string{"status": "ok", "text": text})
}

// SeatAsk proxies seat_ask and announces the single answer.
func (h *Handlers) SeatAsk(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Question string `json:"question"`
		Seat     string `json:"seat"`
		Voice    string `json:"voice"`
		Method   string `json:"method"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		jsonErr(w, 400, "invalid json")
		return
	}
	if strings.TrimSpace(body.Question) == "" {
		jsonErr(w, 400, "question is required")
		return
	}
	text, err := h.rpcCallAs(r, "seat_ask", map[string]any{
		"question": body.Question, "seat": body.Seat,
		"voice": body.Voice, "method": body.Method})
	if err != nil {
		jsonErr(w, 502, err.Error())
		return
	}
	h.broadcast("seat_answered", map[string]string{"text": text})
	jsonResp(w, map[string]string{"status": "ok", "text": text})
}
